dissect/hypervisor/disk/hdd.py

```diff
@@ -0,0 +1,404 @@
+        """
+        for image in self.images:
+            if image.guid == guid:
+                return image
+
+        raise KeyError(f"Image GUID not found: {guid}")
+
+
+@dataclass
+class Image:
+    guid: UUID
+    type: str
+    file: str
+
+    @classmethod
+    def from_xml(cls, element: Element) -> Image:
+        if element.tag != "Image":
+            raise ValueError("Invalid Image XML element")
+
+        return cls(
+            UUID(element.find("GUID").text),
+            element.find("Type").text,
+            element.find("File").text,
+        )
+
+
+@dataclass
+class Snapshots:
+    top_guid: Optional[UUID]
+    shots: list[Shot]
+
+    @classmethod
+    def from_xml(cls, element: Element) -> Snapshots:
+        if element.tag != "Snapshots":
+            raise ValueError("Invalid Snapshots XML element")
+
+        top_guid = element.find("TopGUID")
+        if top_guid:
+            top_guid = UUID(top_guid.text)
+        shots = list(map(Shot.from_xml, element.iterfind("Shot")))
+
+        return cls(top_guid, shots)
+
+    def find_shot(self, guid: UUID) -> Shot:
+        """Find a specific snapshot GUID.
+
+        Args:
+            guid: The snapshot GUID to find.
+
+        Raises:
+            KeyError: If the GUID could not be found.
+        """
+        for shot in self.shots:
+            if shot.guid == guid:
+                return shot
+
+        raise KeyError(f"Shot GUID not found: {guid}")
+
+
+@dataclass
+class Shot:
+    guid: UUID
+    parent: UUID
+
+    @classmethod
+    def from_xml(cls, element: Element) -> Shot:
+        if element.tag != "Shot":
+            raise ValueError("Invalid Shot XML element")
+
+        return cls(
+            UUID(element.find("GUID").text),
+            UUID(element.find("ParentGUID").text),
+        )
+
+
+class StorageStream(AlignedStream):
+    """Stream implementation for HDD streams.
+
+    HDD files can exist of one or multiple streams, starting at consecutive offsets.
+    This class stitches all streams together into a single stream.
+
+    Args:
+        streams: A list of :class:`Storage` and file-like object tuples.
+    """
+
+    def __init__(self, streams: list[tuple[Storage, BinaryIO]]):
+        self.streams = sorted(streams, key=lambda entry: entry[0].start)
+        self._lookup = []
+
+        size = 0
+        for storage, _ in self.streams:
+            self._lookup.append(storage.start)
+            size = storage.end
+
+        super().__init__(size * SECTOR_SIZE)
+
+    def _read(self, offset: int, length: int) -> bytes:
+        sector = offset // SECTOR_SIZE
+        count = (length + SECTOR_SIZE - 1) // SECTOR_SIZE
+
+        result = []
+        stream_idx = bisect_right(self._lookup, sector) - 1
+
+        while count > 0 and stream_idx < len(self.streams):
+            storage, stream = self.streams[stream_idx]
+            sectors_remaining = storage.end - sector
+            read_sectors = min(sectors_remaining, count)
```

Comment:
Does sectors_remaining have the correct value here, in my calculation it is still the raw offset, I do not see where you turn it into sectors instead of bytes? But maybe I am overlooking something here?

Reply:
![image](https://user-images.githubusercontent.com/110607403/232753286-d4618994-8915-409c-9d5c-4c84cb6cc836.png)

I think this is not what it is supposed to be....?

Reply:
It should be? `storage.end` is already in sectors, and both `sector` and `count` are converted to sector numbers at the beginning of this method.

Reply:
Ow ok, it is already in sectors in the xml, check.